Add lint rule for inaccessible disabled `Button`

mirka opened the pull request:
Alternative to #59518
Part of #56547

## What?

Adds an eslint error for when the `disabled` prop is used without the `__experimentalIsFocusable` prop in the `Button` component.

## Why?

We have long been discussing how to prevent accessibility issues stemming from `disabled` controls (#56547). Although there are [situations](https://www.w3.org/WAI/ARIA/apg/practices/keyboard-interface/#focusabilityofdisabledcontrols) where you do want to completely disable a control, we more often find cases where this was not a conscious decision but simply not considered at all. In these cases, we often end up with inaccessible disabled controls, which are not perceivable by screen readers, or cause focus loss after closing a popover.

### Next steps

- Stabilize the `__experimentalIsFocusable` prop. We should consider renaming it to `accessibleWhenDisabled`, which is clear and matches the naming in Ariakit. (#62282)
- Add a `accessibleWhenDisabled` prop to other relevant components, and add those components to the eslint rule.

## How?

Other strategies considered include:

- Changing the behavior of the `disabled` prop to be accessible disabled by default. This is not ideal because it goes against the standard behavior of the `disabled` HTML attribute.
- Changing the name of the `disabled` prop to `isDisabled`, and then make that accessible disabled by default. This is not ideal because it will require a lot of API changes, and the behavioral difference between the two props would be unclear.
- Adding a blanket lint rule for `disabled`, and nudge people toward using `aria-disabled` instead. This is [not ideal](https://github.com/WordPress/gutenberg/pull/59518#issuecomment-1977130497) because using `aria-disabled` is not necessarily going to address all issues, depending on the element/component.
- [Automatically enable](https://github.com/WordPress/gutenberg/issues/56547#issuecomment-1895725290) accessible disabled when a control is dynamically disabled while focused. This is not ideal because it [isn't foolproof](https://github.com/WordPress/gutenberg/pull/59518#issuecomment-2090525451), and also does not address the screen reader perceivability problem.

## Testing Instructions

✅ The linter should error in places where a `Button` has a `disabled` prop without a `__experimentalIsFocusable` prop.
✅ The error message should be clear and actionable.

mirka committed:
Add lint rule for inaccessible disabled `Button`

mirka committed:
Exclude react native files

mirka commented:
When you run `npm run lint:js -- --quiet`, we can see that there are currently 26 violations of this error.

Is anyone available to do a quick sweep of these and fix or ignore as appropriate? If not, I can go in and ignore them with some kind of TODO comment to fix when possible.

tyxla commented on .eslintrc.js:
```diff
@@ -249,6 +249,21 @@ module.exports = {
+					'error',
+					{
+						selector:
+							'JSXOpeningElement[name.name="Button"]:not(:has(JSXAttribute[name.name="__experimentalIsFocusable"])) JSXAttribute[name.name="disabled"]',
```
Confirmed that the rule works well 👍

tyxla commented on .eslintrc.js:
```diff
@@ -249,6 +249,21 @@ module.exports = {
 			},
 		},
+		{
+			files: [ 'packages/*/src/**/*.[tj]s?(x)' ],
```
Should we also enable this for the `storybook` root dir, which may contain some relevant usages?

Example:
https://github.com/WordPress/gutenberg/blob/77a8b55446c7a92ebdd430f2f09eaa804c32f66b/storybook/stories/playground/with-undo-redo/index.js#L51

mirka replied:
Yes, good catch!

tyxla reviewed (commented):
Nice. Haven't approved yet as we still need to deal with the failures.

DaniGuardiola commented:
FWIW, Ariakit handles a lot of this out of the box. I feel like before doing something like this we should just consider re-writing the internals to use Ariakit, keeping the same API. Then, it should be easier to figure out a strategy to nudge users to keep things accessible. This includes some ideas that I have explored in the past like making buttons with tooltips or binded as popover disclosure triggers automatically `accessibleWhenDisabled`, which could take care of most, if not all of this issue, while remaining simpler.

mirka commented:
> FWIW, Ariakit handles a lot of this out of the box. I feel like before doing something like this we should just consider re-writing the internals to use Ariakit, keeping the same API. Then, it should be easier to figure out a strategy to nudge users to keep things accessible. This includes some ideas that I have explored in the past like making buttons with tooltips or binded as popover disclosure triggers automatically `accessibleWhenDisabled`, which could take care of most, if not all of this issue, while remaining simpler.

I have [suggested](https://github.com/WordPress/gutenberg/issues/56547#issuecomment-1895725290) this type of automatic mitigation as a preferred first step as well, but apparently there have already been a couple cases where this would not have been sufficient. Your idea of targeting all buttons bound as popover triggers sounds good, and would work in certain integrated components, but probably not all possible instances of a popover trigger.

My feeling now is that these linting and automating strategies are not mutually exclusive, and would compliment each other. We should continue to look into automating strategies, since that would benefit all consumers and not just Gutenberg where this lint rule would be in effect.

People are already growing weary of the repeated accessibility issues that this introduces, so I'm hoping we can at least prevent new problematic instances and improving developer awareness, while working on some automated strategies in parallel. If they work well enough, maybe we can remove the lint rule. Does that sound reasonable?

mirka committed:
Include files in root `storybook` folder

mirka committed:
Fix in Storybook editor playground (matches actual behavior)

mirka committed:
Fix in install block button (is clearly a busy state)

mirka committed:
Ignore in gallery image reordering buttons

mirka committed:
Fix in LinkControl copy link button (aleviates confusion)

mirka committed:
Ignore in edit-site pagination buttons (not confusing, and useful)

mirka committed:
Fix in enable custom fields (is clearly a busy state)

mirka committed:
Fix in DataViews list view

Empty action menu should still be perceivable to aleviate confusion, and does not clutter tab order due to Composite use.

mirka committed:
Fix in DataViews CompactItemActions (is dropdown trigger)

mirka committed:
Fix in template part title modal

disabled and aria-disabled are set with an identical condition, which doesn't make sense but signals the intent to keep it focusable.

mirka committed:
Fix in PageList block (should be perceivable, esp. because the description is always there)

mirka committed:
Fix in ConvertToLinksModal button (should be perceivable, and doesn't clutter tab order)

mirka committed:
Fix in edit-site "Apply globally" button (should be perceivable, and doesn't clutter tab order)

mirka committed:
Fix in edit-site nav menu rename modal (should be perceivable to signal that input is invalid)

mirka committed:
Fix in RevisionsButtons (button is never visible when `areStylesEqual`)

mirka committed:
Fix in RevisionsButtons (contains important info and should be perceivable)

mirka committed:
Fix in GlobalStylesSidebar (should be perceivable)

mirka committed:
Fix in PostPublishPanel cancel button (can cause focus loss)

mirka committed:
Fix in PostPreviewButton (should be perceivable)

mirka committed:
Defer decision in ButtonBlockAppender

mirka committed:
Fix in reusable blocks import form (should be perceivable, can cause focus loss)

mirka committed:
Adapt test for PostPreviewButton

mirka commented on packages/editor/src/components/post-preview-button/test/index.js:
```diff
@@ -144,7 +144,10 @@ describe( 'PostPreviewButton', () => {
 		render( <PostPreviewButton /> );
 
-		expect( screen.getByRole( 'button' ) ).toBeDisabled();
+		expect( screen.getByRole( 'button' ) ).toHaveAttribute(
+			'aria-disabled',
+			'true'
+		);
```
This is the recommended way to test for `aria-disabled` (see https://github.com/testing-library/jest-dom/issues/144#issuecomment-577235097).

tyxla replied:
Or maybe we can keep both assertions as they check for different things?

mirka replied:
That makes a lot of sense, actually. Addressed in bca37a2.

mirka commented on packages/block-directory/src/plugins/get-install-missing/install-button.js:
```diff
@@ -42,6 +42,7 @@ export default function InstallButton( { attributes, block, clientId } ) {
 					}
 				} )
 			}
+			__experimentalIsFocusable
```
Clearly a busy state, which can cause focus loss.

mirka commented on packages/block-editor/src/components/button-block-appender/index.js:
```diff
@@ -60,6 +60,8 @@ function ButtonBlockAppender(
 						onClick={ onToggle }
 						aria-haspopup={ isToggleButton ? 'true' : undefined }
 						aria-expanded={ isToggleButton ? isOpen : undefined }
+						// TODO: Investigate whether this button should be accessible when disabled.
+						// eslint-disable-next-line no-restricted-syntax
```
This was the only instance I didn't feel confident enough to make a decision without deeper investigation. This is a publicly exported component with both a `tabIndex` and `disabled` prop exposed. (But no usages of these props within the Gutenberg repo.)

tyxla replied:
If I did the archeology correctly, it was added here: https://github.com/WordPress/gutenberg/pull/9815

It can now be seen when inserting buttons in the `Buttons` block:

![Screenshot 2024-05-31 at 14 03 26](https://github.com/WordPress/gutenberg/assets/8436925/0cc6824d-4795-42e4-ac09-6badaf51c749)

Judging by the functionality and the initial intent, my expectation is that this `disabled` was left there unintentionally and there's no reason for it to be there in the first place. I expect that the inserter appender will not be visible at all, rather than stay visible and disabled. With that in mind, I don't think it needs to be accessible if it ends up disabled.

mirka replied:
Thank you so much for investigating! I added a disable reason in d16d2b4.

mirka commented on packages/block-editor/src/components/link-control/link-preview.js:
```diff
@@ -149,6 +149,7 @@ export default function LinkPreview( {
 						isEmptyURL || showIconLabels ? '' : ': ' + value.url
 					) }
 					ref={ ref }
+					__experimentalIsFocusable
```
Alleviates confusion, since this is a "standard" button usually present in the view.

mirka commented on packages/block-library/src/gallery/v1/gallery-image.js:
```diff
@@ -222,13 +222,17 @@ class GalleryImage extends Component {
 						onClick={ isFirstItem ? undefined : onMoveBackward }
 						label={ __( 'Move image backward' ) }
 						aria-disabled={ isFirstItem }
+						// Disable reason: Truly disable when image is not selected.
+						// eslint-disable-next-line no-restricted-syntax
 						disabled={ ! isSelected }
 					/>
 					<Button
 						icon={ chevronRight }
 						onClick={ isLastItem ? undefined : onMoveForward }
 						label={ __( 'Move image forward' ) }
 						aria-disabled={ isLastItem }
+						// Disable reason: Truly disable when image is not selected.
+						// eslint-disable-next-line no-restricted-syntax
 						disabled={ ! isSelected }
```
I was wondering whether a component having both `disabled` and `aria-disabled` props was something to flag with another lint, but this I think is a legitimate use case. (Though the same logic can also be expressed with a combination of the `disabled` and the `__experimentalIsFocusable` prop.)

mirka commented on packages/block-library/src/page-list/convert-to-links-modal.js:
```diff
@@ -27,6 +27,7 @@ export function ConvertToLinksModal( { onClick, onClose, disabled } ) {
 				</Button>
 				<Button
 					variant="primary"
+					__experimentalIsFocusable
```
Should be perceivable, and doesn't clutter tab sequence.

mirka commented on packages/block-library/src/page-list/edit.js:
```diff
@@ -338,6 +338,7 @@ export default function PageListEdit( {
 						<p>{ convertDescription }</p>
 						<Button
 							variant="primary"
+							__experimentalIsFocusable
```
Should be perceivable, especially because the description is always present.

mirka commented on packages/block-library/src/template-part/edit/title-modal.js:
```diff
@@ -43,8 +43,8 @@ export default function TitleModal( { areaLabel, onClose, onSubmit } ) {
 							type="submit"
+							__experimentalIsFocusable
 							disabled={ ! title.length }
-							aria-disabled={ ! title.length }
```
`disabled` and `aria-disabled` were set with an identical condition, which is clearly a mistake, but signals the intent to keep it focusable.

mirka commented on packages/dataviews/src/item-actions.tsx:
```diff
@@ -254,6 +254,7 @@ function CompactItemActions< Item extends AnyItem >( {
 					size="compact"
 					icon={ moreVertical }
 					label={ __( 'Actions' ) }
+					__experimentalIsFocusable
```
This is a dropdown trigger, so it can potentially cause focus loss.

mirka commented on packages/dataviews/src/view-list.tsx:
```diff
@@ -255,6 +255,7 @@ function ListItem< Item extends AnyItem >( {
 												size="compact"
 												icon={ moreVertical }
 												label={ __( 'Actions' ) }
+												__experimentalIsFocusable
```
An empty action menu should still be perceivable to alleviate confusion, and does not clutter tab sequence due to `Composite` use.

mirka commented on packages/edit-post/src/components/preferences-modal/enable-custom-fields.js:
```diff
@@ -42,6 +42,7 @@ export function CustomFieldsConfirmation( { willEnable } ) {
 				className="edit-post-preferences-modal__custom-fields-confirmation-button"
 				variant="secondary"
 				isBusy={ isReloading }
+				__experimentalIsFocusable
```
Clearly a busy state, which can cause focus loss.

mirka commented on packages/edit-site/src/components/global-styles-sidebar/index.js:
Both buttons should be perceivable, because they are expected to be there.

mirka commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -219,7 +220,6 @@ function RevisionsButtons( {
 								</p>
 							) : (
 								<Button
-									disabled={ areStylesEqual }
```
Due to the rendering logic, this button will never be shown when `areStylesEqual` is truthy. (⚠️ Please double-check that I'm reading this correctly!)

tyxla replied:
Confirmed that this can and should be removed because `areStylesEqual` will never be true since it's already `false` based on the ternary logic this is nested in.

mirka commented on packages/edit-site/src/components/global-styles/screen-revisions/revisions-buttons.js:
```diff
@@ -163,6 +163,7 @@ function RevisionsButtons( {
 					>
 						<Button
 							className="edit-site-global-styles-screen-revisions__revision-button"
+							__experimentalIsFocusable
```
This button can contain a lot of important information and should be perceivable even when disabled.

mirka commented on packages/edit-site/src/components/pagination/index.js:
These pagination buttons wouldn't cause confusion when truly disabled, and removing irrelevant buttons from the tab sequence would help a keyboard user get to the nav button they want quicker.

mirka commented on packages/edit-site/src/components/sidebar-navigation-screen-navigation-menu/rename-modal.js:
```diff
@@ -43,6 +43,7 @@ export default function RenameModal( { menuTitle, onClose, onSave } ) {
 
 						<Button
 							__next40pxDefaultSize
+							__experimentalIsFocusable
```
Should be perceivable to signal that the user input is invalid.

mirka commented on packages/editor/src/components/post-preview-button/index.js:
```diff
@@ -168,6 +168,7 @@ export default function PostPreviewButton( {
 			className={ className || 'editor-post-preview' }
 			href={ href }
 			target={ targetId }
+			__experimentalIsFocusable
```
Should be perceivable, because the button is expected to be there.

mirka commented on packages/editor/src/components/post-publish-panel/index.js:
```diff
@@ -93,6 +93,7 @@ export class PostPublishPanel extends Component {
 							</div>
 							<div className="editor-post-publish-panel__header-cancel-button">
 								<Button
+									__experimentalIsFocusable
```
Can cause focus loss due to a dynamic disable.

mirka commented on packages/list-reusable-blocks/src/components/import-form/index.js:
```diff
@@ -86,6 +86,7 @@ function ImportForm( { instanceId, onUpload } ) {
 			<Button
 				type="submit"
 				isBusy={ isLoading }
+				__experimentalIsFocusable
```
Should be perceivable, and can cause focus loss.

mirka commented on storybook/stories/playground/with-undo-redo/index.js:
Accessible disabled matches the behavior of the undo/redo buttons in the actual WP block editor.

mirka reviewed (commented):
I went ahead and looked through all of the errors myself, since I realized it would help me get a better overview of our problems.

For most of them I made a decision based on looking at the code only. Some of them, I looked into the original PR, blames, or actual app UI to gather more context. None of them are _tested_ in app. I believe this is sufficient because technically this shouldn't be introducing "regressions" — there is little damage that adding a `__experimentalIsFocusable` could do. The only potentially regressive change is https://github.com/WordPress/gutenberg/pull/62080#discussion_r1621385134, but the logic looks clear enough to me.

I wrote brief reasoning comments for each change to leave a trail in case somebody needs to `blame`.

mirka commented:
> My feeling now is that these linting and automating strategies are not mutually exclusive, and would complement each other.

After going through all these errors manually, I am now more convinced that we should be approaching this with a combination of automated mitigation and human decision-making (prompted by lints). Automated mitigation can prevent a baseline number of accessibility issues, but a human making context-aware design decisions for each instance can further enhance the UX for screen reader users.

tyxla reviewed (approved):
Thanks for taking the time to audit and fix them one by one! 👍 

I spent some time going through them all with your comments and I think they all make sense. I've left a few comments where there was doubt or uncertainty.

I think it can be shipped - do you think we should ask folks from #59518 for some feedback just in case?

mirka committed:
Improve rigidity of accessible disabled detection in test

mirka committed:
Add disable reason for ButtonBlockAppender

mirka commented:
> I think it can be shipped - do you think we should ask folks from #59518 for some feedback just in case?

Thanks for the review, I appreciate it 🙏 

Since this touches many files, I'll go ahead and merge today. It should be easy enough to revert or tweak, whether in full or in part.

afercia commented:
Thanks everyone for all your effort ❤️  A linting rule seems the best approach to prevent such a11y issues and, more importantly, to improve knowledge and education.
If needed, all the cases investigated here can be tweaked but it's very good there's now a solid base to catch violations.

ellatrix commented:
This is a bug fix no? Looks like a bunch of buttons were made more accessible.

ellatrix commented:
This was cherry-picked to the wp/6.6 branch.
